components/browser/menu/src/main/java/mozilla/components/browser/menu/item/CustomTooltip.kt

```diff
@@ -0,0 +1,152 @@
+        // Measure pop-up
+        val spec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
+        popupWindow?.contentView?.measure(spec, spec)
+
+        val rootView = anchor.rootView
+        val rootPosition = IntArray(2)
+        val anchorPosition = IntArray(2)
+        rootView.getLocationOnScreen(rootPosition)
+        anchor.getLocationOnScreen(anchorPosition)
+
+        val rootY = rootPosition[1]
+        val anchorY = anchorPosition[1]
+
+        val rootHeight = rootView.height
+        val tooltipHeight = popupWindow?.contentView?.measuredHeight ?: 0
+        val tooltipWidth = popupWindow?.contentView?.measuredWidth ?: 0
+
+        val checkY = rootY + rootHeight - (anchorY + anchor.height + tooltipHeight + TOOLTIP_EXTRA_VERTICAL_OFFSET_DP)
+        val belowY = TOOLTIP_EXTRA_VERTICAL_OFFSET_DP
+        val aboveY = -(anchor.height + tooltipHeight + TOOLTIP_EXTRA_VERTICAL_OFFSET_DP)
+
+        // align anchor center with tooltip center
+        val offsetX = anchor.width / 2 - tooltipWidth / 2
+        // make sure tooltip is visible and it's not displayed below, outside the view
+        val offsetY = if (checkY > 0) belowY else aboveY
+        return Offset(offsetX, offsetY)
+    }
+
+    @SuppressLint("InflateParams")
+    fun show() {
+        activeTooltip?.hide()
+        activeTooltip = this
+
+        val layout = LayoutInflater.from(anchor.context)
+            .inflate(R.layout.mozac_browser_tooltip_layout, null)
+
+        layout.findViewById<TextView>(R.id.mozac_browser_tooltip_text).text = tooltipText
+        popupWindow = PopupWindow(
+            layout,
+            LinearLayout.LayoutParams.WRAP_CONTENT,
+            LinearLayout.LayoutParams.WRAP_CONTENT,
+            false
+        )
+
+        val offsets = computeOffsets()
+
+        popupWindow?.let {
+            PopupWindowCompat.setWindowLayoutType(it, WindowManager.LayoutParams.TYPE_APPLICATION_SUB_PANEL)
+            it.isTouchable = false
+            it.showAsDropDown(anchor, offsets.x, offsets.y, Gravity.CENTER)
+        }
+
+        anchor.removeCallbacks(hideTooltipRunnable)
+        anchor.postDelayed(hideTooltipRunnable, LONG_CLICK_HIDE_TIMEOUT_MS)
+    }
+
+    fun hide() {
+        if (activeTooltip === this) {
+            activeTooltip = null
+            popupWindow?.let {
+                it.dismiss()
+                popupWindow = null
+                anchor.removeOnAttachStateChangeListener(this)
+            }
+        }
+    }
+
+    override fun onViewAttachedToWindow(v: View) {
+        // no-op
+    }
+
+    override fun onViewDetachedFromWindow(v: View) {
+        hide()
+        anchor.removeCallbacks(hideTooltipRunnable)
+    }
+
+    companion object {
+        private const val LONG_CLICK_HIDE_TIMEOUT_MS: Long = 2500
+        private const val TOOLTIP_EXTRA_VERTICAL_OFFSET_DP = 8
+        // The tooltip currently being shown
+        private var activeTooltip: CustomTooltip? = null
```

Comment:
Let's address this check
```
[task 2020-05-04T16:33:07.856Z]   Errors found:
[task 2020-05-04T16:33:07.856Z]   
[task 2020-05-04T16:33:07.856Z]   /builds/worker/checkouts/src/components/browser/menu/src/main/java/mozilla/components/browser/menu/item/CustomTooltip.kt:124: Error: Do not place Android context classes in static fields (static reference to CustomTooltip which has field anchor pointing to View); this is a memory leak [StaticFieldLeak]
[task 2020-05-04T16:33:07.856Z]           // The tooltip currently being shown
[task 2020-05-04T16:33:07.856Z]           ^
[task 2020-05-04T16:33:07.856Z]   
[task 2020-05-04T16:33:07.856Z] 
```

Reply:
@Amejia481 I cannot replicate the exact behavior of a tooltip without using using a static reference to the _active_ (currently visible) tooltip. 

The options are either to suppress this warning or slightly change the behavior to allow for the tooltip to disappear when you click outside it (a normal tooltip does not disappear until the timeout has passed, another tooltip is triggered to be shown or the anchor view is detached from window). It's not a big difference IMO. What do you think would be the best option?

Reply:
Got it, we can suppress it, but let's make sure we are cleaning `activeTooltip` after is not needed any more to avoid memory leaks 👍

Reply:
Thanks for the review! I added the suppress annotation and explanation in comment. I have investigated with both Leak Canary and A-S profiler and they showed no memory leaks.

Reply:
Awesome thank you for the great work!